### PR TITLE
GenericFlinkSource - source can define more complex behaviour

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala [hunk 1]
```diff
@@ -1,17 +1,28 @@
...
+import pl.touk.nussknacker.engine.api.{MetaData, MethodToInvoke}
 import pl.touk.nussknacker.engine.api.process.{Source, SourceFactory}
-import pl.touk.nussknacker.engine.api.test.TestDataParser
 
 /**
   * Source with methods specific for Flink
   * @tparam T - type of event that is generated by this source. This is needed to handle e.g. syntax suggestions in UI
   */
-trait FlinkSource[T] extends Source[T] {
+trait BaseFlinkSource[T] extends Source[T] {
```
Comment (arkadius) [inline engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala]: Please swap names: FlinkSource <> BaseFlinkSource and add the same abstraction layers for tests

File: engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala [hunk 2]
```diff
@@ -20,8 +31,23 @@ trait FlinkSource[T] extends Source[T] {
 
   def timestampAssigner : Option[TimestampAssigner[T]]
 
-}
+  override def startStream(env: StreamExecutionEnvironment, metaData: MetaData): DataStream[T] = {
```
Comment (arkadius) [inline engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala]: WDYT about rename it to `sourceStream`? WDYT about extracting metaData to implicit parameters - like it is in other places in our API?

Comment (mproch) [reply engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala]: done :+1:

File: engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala (hunk 1 above)
Comment (mproch) [reply engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala]: :+1:  I swapped names, tests are a bit clumsy for now but I don't have really elegant idea for now..

Comment (arkadius) [reply engine/flink/api/src/main/scala/pl/touk/nussknacker/engine/flink/api/process/FlinkSource.scala]: Looks good. If you already have changed this api, can you also rename `toFlinkSource`->`flinkSource` to clarify that objects of this class are not converted to flink source but flink source is used for preparing DataStream?

**Pull request merged**